ext/socket/lib/socket.rb

```diff
@@ -623,45 +684,131 @@ def accept_nonblock(exception: true)
+    getaddrinfo_v4_done = attempt_v4 ? false : true
+    while ai = queue.pop do
+      break if queue.closed? # timed out
+      case ai
+      when :getaddrinfo_v6_done
+        getaddrinfo_v6_done = true
+      when :getaddrinfo_v4_done
+        getaddrinfo_v4_done = true
       else
         local_addr = nil
+        if local_addr_list
+          local_addr = local_addr_list.find {|local_ai| local_ai.afamily == ai.afamily }
+          next unless local_addr
+        end
+        begin
+          sock = Socket.new(ai.pfamily, ai.socktype, ai.protocol)
+          sock.ipv6only! if ai.ipv6?
+          sock.bind(local_addr) if local_addr
+          case sock.connect_nonblock(ai, exception: false)
+          when 0
+            ret = sock
+            break
+          when :wait_writable
+            socket_list.push(sock)
+            # 250ms is the recommended value for the connection attempt delay in RFC8305
+            _, writable_sockets, = IO.select(nil, socket_list, nil, 0.25)
+            ret, error = find_connected_socket(writable_sockets)
+            last_error = error if error
+            break if ret
+          end
+        rescue SystemCallError => ex
+          last_error = ex
+        end
+      end
+      if getaddrinfo_v6_done && getaddrinfo_v4_done && queue.empty?
+        break # all connection attempts has been made
       end
+    end
+
+    # wait for sockets to be connected
+    if !ret && !socket_list.empty?
       begin
-        sock = local_addr ?
-          ai.connect_from(local_addr, timeout: connect_timeout) :
-          ai.connect(timeout: connect_timeout)
-      rescue SystemCallError
-        last_error = $!
-        next
+        if connect_timeout
+          if queue.closed? # timed out
+            timeout = 0 # returns immediately
+          else
+            # wait until connect_timeout
+            elapsed = Process.clock_gettime(Process::CLOCK_MONOTONIC) - start_time
+            timeout = connect_timeout - elapsed
+            timeout = 0 if timeout.negative?
+          end
+        else
+          timeout = nil # wait forever
+        end
+        _, writable_sockets, = IO.select(nil, socket_list, nil, timeout)
+      rescue SystemCallError => ex
+        last_error = ex
       end
-      ret = sock
-      break
-    }
+      ret, error = find_connected_socket(writable_sockets)
+      last_error = error if error
+    end
+
     unless ret
       if last_error
         raise last_error
-      else
+      elsif connect_timeout
+        raise Errno::ETIMEDOUT, "user specified timeout"
+      elsif local_addr_list
         raise SocketError, "no appropriate local address"
       end
     end
+
+    ret.nonblock = false
+
     if block_given?
       begin
         yield ret
       ensure
         ret.close
       end
     else
-      ret
+      return ret
+    end
+  ensure
+    # cleanup
+    getaddrinfo_v6_th&.exit
+    getaddrinfo_v4_th&.exit
+    timeout_th&.exit
+    socket_list&.each do |socket|
```

Comment:
Just a tiny nit, but looks like `socket_list` might be empty but won't be `nil` so I think no need for `&` on this one.

Reply:
An exception might be raised before `socket_list` is assigned.
That's why we need `&` in this case.